nimble/host/services/cts/src/ble_svc_cts.c

```diff
@@ -0,0 +1,544 @@
+
+    /* Copy data */
+    if (os_mbuf_append(ctxt->om, &ble_svc_cts_local_time_information,
+		                 sizeof(ble_svc_cts_local_time_information))) {
+	return BLE_ATT_ERR_INSUFFICIENT_RES;
+    }
+    
+    return 0;
+}
+
+#if MYNEWT_VAL(BLE_SVC_CTS_LOCAL_TIME_INFORMATION_WRITE_PERM) >= 0
+static inline int
+ble_svc_cts_local_time_information_write_access(struct ble_gatt_access_ctxt *ctxt)
+{
+    /* Retrieve data */
+    uint16_t om_len = OS_MBUF_PKTLEN(ctxt->om);
+    if (om_len > sizeof(ble_svc_cts_local_time_information)) {
+        return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
+    }
+    if (ble_hs_mbuf_to_flat(ctxt->om,
+			    &ble_svc_cts_local_time_information, om_len, NULL)){
+        return BLE_ATT_ERR_UNLIKELY;
+    }
+
+    /* Get current timezone/dst */
+    struct os_timezone tz_o;
+    if (os_gettimeofday(NULL, &tz_o)) {
+	return BLE_ATT_ERR_UNLIKELY;
+    }
+    
+    /* Compute new timezone/dst */
+    struct os_timezone tz_n = {
+	.tz_minuteswest = ble_svc_cts_local_time_information.time_zone * 15,
+	.tz_dsttime     = ble_svc_cts_local_time_information.dst_offset != 0
+    };
+
+    /* Ensure conversion is legit 
+     * For DST, see: https://github.com/apache/mynewt-core/issues/944
+     */
+    if (ble_svc_cts_local_time_information.dst_offset != 4) {
+        return BLE_ATT_ERR_UNLIKELY; /* XXX: better error code? */
+    }
+	
+    /* Set new timezone/dst */
+    if (os_settimeofday(NULL, &tz_n)) 
+	return BLE_ATT_ERR_UNLIKELY;
+
+#if MYNEWT_VAL(BLE_SVC_CTS_CURRENT_TIME_NOTIFY_ENABLE) > 0
+    /* Notify */
+    ble_svc_cts_adjust_reason = 0;
+    if (tz_o.tz_minuteswest != tz_n.tz_minuteswest) {
+	ble_svc_cts_adjust_reason |=
+	    BLE_SVC_CTS_ADJUST_REASON_CHANGE_OF_TIMEZONE;
+    }
+    if ((tz_o.tz_dsttime != 0) != (tz_n.tz_dsttime != 0)) {
+	ble_svc_cts_adjust_reason |=
+	    BLE_SVC_CTS_ADJUST_REASON_CHANGE_OF_DST;
+    }
+    ble_gattc_notify(ble_svc_cts_conn_handle,
+		     ble_svc_cts_current_time_handle);
+#endif
+    
+    return 0;
+}
+#endif
+#endif
+
+#if MYNEWT_VAL(BLE_SVC_CTS_REFERENCE_TIME_INFORMATION_READ_PERM) >= 0
+static inline int
+ble_svc_cts_reference_time_information_read_access(struct ble_gatt_access_ctxt *ctxt)
+{
+    int rc;
+
+    /* Check that reference time has been used at least once to update
+     * time, otherwise use the initial value, which is unknown 
+     */
+    if (ble_svc_cts_reference_time_was_used) {
+	/* Get current time */
+	struct os_timeval  tv;
+	struct os_timezone tz;
+	if (os_gettimeofday(&tv, &tz)) {
+	    return BLE_ATT_ERR_UNLIKELY;
+	}
+
+	/* Compute delta since last update */
+	uint64_t delta = tv.tv_sec - ble_svc_cts_reference_time_last_used; 
+
+	/* Convert to days and hours format */
+	uint8_t  days, hours;
+	delta /= 3600; 
+	hours  = delta % 24;
+	delta /= 24;
+	days   = delta;
+
+	/* Special case if more than 254 days ago */
+	if (delta > 254) {
+	    days = hours = BLE_SVC_CTS_UPDATED_SINCE_A_LONG_TIME_AGO;
```

Comment:
no double assignments

also symbol should have different name, it does not quite fit here imo. I'd make it sth like `BLE_SVC_CTS_UPDATED_255_OR_MORE'